src/pip/_internal/operations/prepare.py

```diff
@@ -498,12 +501,14 @@ def prepare_linked_requirements_more(
         reqs = [req for req in reqs if req.needs_more_preparation]
         for req in reqs:
             # Determine if any of these requirements were already downloaded.
-            if self.download_dir is not None and req.link.is_wheel:
-                hashes = self._get_linked_req_hashes(req)
-                file_path = _check_download_dir(req.link, self.download_dir, hashes)
-                if file_path is not None:
-                    self._downloaded[req.link.url] = file_path
-                    req.needs_more_preparation = False
+            if self.download_dir is not None:
+                assert req.link is not None
+                if req.link.is_wheel:
+                    hashes = self._get_linked_req_hashes(req)
+                    file_path = _check_download_dir(req.link, self.download_dir, hashes)
+                    if file_path is not None:
+                        self._downloaded[req.link.url] = file_path
+                        req.needs_more_preparation = False
```

Comment:
This can be `if self.download_dir is not None and req.link is not None and req.link.is_wheel:` which is semantically slightly different but do the same thing functionally.